### PR TITLE
Refactor sudt with registry address

### PR DESCRIPTION
**Pull request by jjyr**
### Core data structure

* Core data structure changes: https://github.com/nervosnetwork/godwoken/pull/649
* All short script hash is removed, replaced by address registry

#### Files
* `gw_registry_addr.h`, address data structure
* `gw_eth_addr_reg.h`
* `gw_eth_addr_reg.c`
* `crates/state-validator`

#### Changes

* New builtin script: Ethereum address registry.
* User deposit will mapping ethereum address to the Ethereum address registry.

#### Files
* `sudt_utils.h`
* `generator_utils.h`
* `validator_utils.h`

#### Changes

The following new syscall is added:

* sys_get_registry_address_by_script_hash
* sys_get_script_hash_by_registry_address

The following syscall is removed:

* sys_get_script_hash_by_prefix - used to get script_hash by short_script_hash

Short script hash in `sudt_utils.h` is replaced by registry address

#### Files

* `crates/challenge-lock`

#### Changes

* Support signing tx and withdrawal in Ethereum tx format and EIP-712.
* Using `RollupConfig#chain_id` in tx, withdrawal signature.

### CODE REVIEW

File: c/common.h [hunk 1]
```diff
@@ -252,32 +233,92 @@ int _check_data_hash_exist(gw_context_t *ctx, uint8_t data_hash[32],
...
+    printf(
+        "_gw_build_registry_address_to_script_hash_key: invalid addr size, "
+        "expect 28");
+    return GW_FATAL_BUFFER_OVERFLOW;
   }
-  /* Check short_script_hash_key */
-  uint8_t raw_key[GW_KEY_BYTES] = {0};
-  int ret = gw_build_short_script_hash_to_script_hash_key(
-      short_script_hash, short_script_hash_len, raw_key);
+  memcpy(key, (uint8_t *)"reg", 3);
```
Comment (Flouse) [inline c/common.h:256]:
```suggestion

  /* raw_key 32 bytes = 3 + 1 + 4 + 4 + 20 */
  memcpy(key, (uint8_t *)"reg", 3);
```

File: c/common.h [hunk 2]
```diff
@@ -252,32 +233,92 @@ int _check_data_hash_exist(gw_context_t *ctx, uint8_t data_hash[32],
...
+  memcpy(key + 4, script_hash, 32);
+}
+
+int _gw_build_registry_address_to_script_hash_key(uint8_t key[32],
+                                                  gw_reg_addr_t *addr) {
+  /* format: "reg" | flag(1 byte) | registry_address
+  registry_address: registry_id(4 bytes) | address_len(4 bytes) | address(n
+  bytes) */
+  if (GW_REG_ADDR_SIZE((*addr)) != 28) {
+    /* raw_key 32 bytes = 3 + 1 + 4 + 4 + 20 */
```
Comment (Flouse) [inline c/common.h:250]:
```suggestion
```

File: tests/src/script_tests/l2_scripts/examples.rs [hunk 3]
```diff
@@ -343,14 +339,14 @@ fn test_example_account_operation() {
     // Log: success
     {
         let account_id = 0;
-        let from_script_hash = [0x33u8; 32];
-        let to_script_hash = [0x44u8; 32];
+        let registry_id = 2;
```
Comment (Flouse) [inline tests/src/script_tests/l2_scripts/examples.rs:342]:
```suggestion
        let registry_id = gw_common::builtins::ETH_REGISTRY_ACCOUNT_ID;
```

**Review by Flouse — APPROVED** [review]

**Review by zeroqn — APPROVED** [review]

File: c/generator_utils.h [hunk 4]
```diff
@@ -380,9 +367,17 @@ int sys_pay_fee(gw_context_t *ctx, const uint8_t *payer_short_script_hash,
   if (ret != 0) {
     return ret;
   }
+  uint8_t buf[32] = {0};
+  int len = GW_REG_ADDR_SIZE(addr);
+  if (len > 32) {
```
[the remark this answers is not in the document]
Comment (jjyr) [reply c/generator_utils.h:372]: 32 is the length of buffer, actually this function allows an address larger than 28. The only limitation is the buffer length.

**Pull request merged**